File: 3470.permutations-iv.py
#
# @lc app=leetcode id=3470 lang=python3
#
# [3470] Permutations IV
#

# @lc code=start
class Solution:
    def permute(self, n: int, k: int) -> List[int]:
        dp = [[[0] * 2 for _ in range(51)] for _ in range(51)]
        dp[0][0][0] = 1
        dp[0][0][1] = 1
        for o in range(51):
            for e in range(51):
                if o == 0 and e == 0:
                    continue
                if o > 0:
                    dp[o][e][1] = o * dp[o - 1][e][0]
                if e > 0:
                    dp[o][e][0] = e * dp[o][e - 1][1]
        num_o = (n + 1) // 2
        num_e = n // 2
        total = dp[num_o][num_e][0] + dp[num_o][num_e][1]
        if k > total:
            return []
        used = set()
        res = []
        remain_o = num_o
        remain_e = num_e
        prev_par = -1
        curr_k = k
        for i in range(n):
            for c in range(1, n + 1):
                if c in used:
                    continue
                par = c % 2
                if i > 0 and par == prev_par:
                    continue
                o_after = remain_o - (1 if par == 1 else 0)
                e_after = remain_e - (1 if par == 0 else 0)
                next_p = 1 - par
                ways = dp[o_after][e_after][next_p]
                if curr_k <= ways:
                    res.append(c)
                    used.add(c)
                    remain_o = o_after
                    remain_e = e_after
                    prev_par = par
                    break
                else:
                    curr_k -= ways
            else:
                return []
        return res

# @lc code=end

File: 3470.permutations-iv.java
#
# @lc app=leetcode id=3470 lang=java
#
# [3470] Permutations IV
#
# @lc code=start
import java.util.*;

class Solution {
    public int[] permute(int n, long k) {
        List<Integer> odds = new ArrayList<>();
        List<Integer> evens = new ArrayList<>();
        for (int i = 1; i <= n; ++i) {
            if (i % 2 == 1) odds.add(i);
            else evens.add(i);
        }
        // dp[remOdd][remEven][lastParity], lastParity: 0=odd, 1=even, 2=start
        Long[][][] dp = new Long[odds.size()+1][evens.size()+1][3];
        // Count all valid alternating perms with given counts and lastParity
        class Counter {
            long count(int ro, int re, int lp) {
                if (ro==0 && re==0) return 1;
                if (dp[ro][re][lp] != null) return dp[ro][re][lp];
                long total = 0;
                if (lp == 2) {
                    // Start: can pick odd or even
                    if (ro > 0) total += ro * count(ro-1, re, 0);
                    if (re > 0) total += re * count(ro, re-1, 1);
                } else if (lp == 0) {
                    // Last was odd, must pick even
                    if (re > 0) total += re * count(ro, re-1, 1);
                } else {
                    // Last was even, must pick odd
                    if (ro > 0) total += ro * count(ro-1, re, 0);
                }
                return dp[ro][re][lp] = total;
            }
        }
        Counter counter = new Counter();
        long total = counter.count(odds.size(), evens.size(), 2);
        if (k > total) return new int[0];
        // Build the permutation step by step
        boolean[] used = new boolean[n+1];
        int[] ans = new int[n];
        int ro = odds.size(), re = evens.size(), lp = 2;
        for (int pos = 0; pos < n; ++pos) {
            boolean found = false;
            for (int cand = 1; cand <= n; ++cand) {
                if (used[cand]) continue;
                int candParity = cand % 2;
                if (lp == 2 || candParity != lp) {
                    int nro = ro, nre = re;
                    if (candParity == 0) nre--; else nro--;
                    if (nro < 0 || nre < 0) continue;
                    long cnt = counter.count(nro, nre, candParity);
                    if (k > cnt) {
                        k -= cnt;
                    } else {
                        ans[pos] = cand;
                        used[cand] = true;
                        ro = nro; re = nre; lp = candParity;
                        found = true;
                        break;
                    }
                }
            }
            if (!found) return new int[0]; // Could not find a valid candidate
        }
        // Verification step: check all constraints
        if (ans.length != n) return new int[0];
        boolean[] checkUsed = new boolean[n+1];
        for (int i = 0; i < n; ++i) {
            if (ans[i] < 1 || ans[i] > n || checkUsed[ans[i]]) return new int[0];
            checkUsed[ans[i]] = true;
            if (i > 0 && (ans[i]%2) == (ans[i-1]%2)) return new int[0];
        }
        return ans;
    }
}
# @lc code=end